src/main/java/com/streamr/client/dataunion/DataUnion.java

```diff
@@ -0,0 +1,233 @@
+package com.streamr.client.dataunion;
+
+import com.streamr.client.dataunion.contracts.DataUnionMainnet;
+import com.streamr.client.dataunion.contracts.DataUnionSidechain;
+import com.streamr.client.dataunion.contracts.IERC20;
+import com.streamr.client.utils.Web3jUtils;
+import org.apache.commons.lang3.reflect.FieldUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.web3j.abi.TypeEncoder;
+import org.web3j.abi.datatypes.*;
+import org.web3j.abi.datatypes.generated.Uint256;
+import org.web3j.crypto.Credentials;
+import org.web3j.crypto.ECKeyPair;
+import org.web3j.crypto.Sign;
+import org.web3j.protocol.Web3j;
+import org.web3j.protocol.core.methods.response.TransactionReceipt;
+import org.web3j.tx.Contract;
+import org.web3j.tx.ManagedTransaction;
+import org.web3j.utils.Numeric;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.math.BigInteger;
+import java.util.ArrayList;
+
+import static com.streamr.client.utils.Web3jUtils.*;
+
+public class DataUnion {
+    private static final Logger log = LoggerFactory.getLogger(DataUnion.class);
+    public enum ActiveStatus{none, active, inactive};
+
+    private DataUnionMainnet mainnet;
+    private DataUnionSidechain sidechain;
+    private Web3j mainnetConnector, sidechainConnector;
+    private Credentials mainnetCred, sidechainCred;
+
+    //use DataUnionClient to instantiate
+    protected DataUnion(DataUnionMainnet mainnet, Web3j mainnetConnector, Credentials mainnetCred, DataUnionSidechain sidechain, Web3j sidechainConnector, Credentials sidechainCred) {
+        this.mainnet = mainnet;
+        this.mainnetConnector = mainnetConnector;
+        this.mainnetCred = mainnetCred;
+        this.sidechain = sidechain;
+        this.sidechainConnector = sidechainConnector;
+        this.sidechainCred = sidechainCred;
+    }
+
+    public boolean waitForDeployment(long pollInterval, long timeout) throws Exception {
+        String code = waitForCodeAtAddress(sidechain.getContractAddress(), sidechainConnector, pollInterval, timeout);
+        return code != null && !code.equals("0x");
+    }
+
+    public String mainnetAddress(){
+        return mainnet.getContractAddress();
+    }
+
+    public String sidechainAddress(){
+        return sidechain.getContractAddress();
+    }
+
+    public boolean isDeployed() throws Exception {
+        //this will check the condition only once:
+        return waitForDeployment(0,0);
+    }
+
+    public void sendTokensToBridge() throws Exception {
+        mainnet.sendTokensToBridge().send();
+    }
+
+    public BigInteger waitForEarningsChange(final BigInteger initialBalance, long pollInterval, long timeout) throws Exception {
+        Web3jUtils.Condition earningsChange = new Web3jUtils.Condition(){
+            @Override
+            public Object check() throws Exception {
+                BigInteger bal = sidechain.totalEarnings().send().getValue();
```

Comment:
Method signature `throws Exception` tells no valuable information for user of this client library.
`sidechain.totalEarnings()` cannot throw an exception. Only `send()` does. `send()`'s `throw Exception` inherits from `RemoteCall`.
Web3j library throws two types of exceptions:
- `IOException` for an error in internet connection. We should catch that and rethrow it as a custom domain specific checked exception. For example: `ConnectionException`.
- `TransactionException` for an error in blockchain transaction. We should also catch and rethrow it as a custom domain specific checked exception. For example: `TransactionException`.

These custom exceptions should be defined in the method signature: `public BigInteger waitForEarningsChange(final BigInteger initialBalance, long pollInterval, long timeout) throws ConnectionException, TransactionException`
and also in `Web3jUtils`:
`public static Object waitForCondition(Condition condition, final long sleeptime, final long timeout) throws ConnectionException, TransactionException`

Reply:
per today's discussion: lets keep throws Exception for now. This is really an issue with Web3j. Best way to solve is to send them PR.